deepchecks/recommender/ranking.py

```diff
@@ -7,705 +7,423 @@
-    item2ix = {}
-    counter = 0
-    for prediction in recommendations:
-        for item in prediction:
-            if item not in item2ix:
-                item2ix[item] = counter
-                counter += 1
+    # Compute the mean NDCG@k and return
+    return mean_ndcg / num_items
 
-    n_items = len(item2ix.keys())
-
-    # create matrix of predicted x items
-    items_matrix = np.zeros((n_predictions, n_items))
-
-    for i, prediction in enumerate(recommendations):
-        for item in prediction:
-            item_ix = item2ix[item]
-            items_matrix[i][item_ix] = 1
-
-    similarity = cosine_similarity(X=items_matrix)
-    dim = similarity.shape[0]
-    personalization = (similarity.sum() - dim) / (dim * (dim - 1))
-
-    return 1 - personalization
-
-
-def novelty(recommendation: List, item_popularity: dict, num_users: int, k: int = 10):
+def popularity_based_novelty(recommendations: Union[List[list], np.array],
+                             items_popularity: Union[dict, None],
+                             num_users: int,
+                             k: int = 20) -> Tuple[float, float]:
     """
-    Calculates the capacity of the recommender system to generate novel
-    and unexpected results.
-
-    Args:
-        recommendation (List): A 1 x N array of items, where M is the number
-                               of predicted lists and N the number of recommended items
-        item_popularity (dict): A dict mapping each item in the recommendations to a popularity value.
-                                  Popular items have higher values.
-        num_users (int): The number of users
-        k (int): The number of items considered in each recommendation.
+    Calculate the popularity-based novelty metric for a list of recommendations.
+    Parameters:
+        recommendations (Union[List[list], np.array]): List of recommendations for each user.
+        items_popularity (Union[dict, None]): Item popularity scores.
+        num_users (int): Total number of users.
+        k (int): The number of top items to consider for novelty. Default is 20.
 
     Returns:
-        (float): novelty
-
-    Metric Definition:
-    Zhou, T., Kuscsik, Z., Liu, J. G., Medo, M., Wakeling, J. R., & Zhang, Y. C. (2010).
-    Solving the apparent diversity-accuracy dilemma of recommender systems.
-    Proceedings of the National Academy of Sciences, 107(10), 4511-4515.
-
-    `Original <https://github.com/statisticianinstilettos/recmetrics/blob/master/recmetrics/metrics.py#L14>`_
-    """
-    if item_popularity is None:
-        raise ValueError("item_popularity must be provided.")
-
-    epsilon = 1e-10
+        Tuple[float, float]: The popularity-based novelty score and the normalized score.
+    """
+    #epsilon = 1e-10
+    # Calculate the mean surprisal/self-information of the top K popular head items
+    top_information_mean = 0
+    most_popular_item_catalog = round(0.05 * len(items_popularity))
```

Comment:
Shouldn't this cutoff (5% most popular items) be a parameter?

Reply:
Good point. Fixed.